Added indel support to SamLocusIterator

magicDGS opened the pull request:
Added indel tracking to `SamLocusIterator` (as discussed in issue #387) and some tests. The default behaviour is still do not keep reads spanning indels.

tfenne commented:
Hi @magicDGS, thanks for doing this!  I'll try and take a look at this in the next day or so.  In the mean time two quick things:
1. It looks like the test that is failing is a new test that you added - do you know what's going on?
2. Your edits introduce a lot of hard-tabs.  Some time before merging those'll need to get turned into soft-tabs with tab size=4.

magicDGS commented:
Sorry, I forgot to add some changes in my local repository before PR. It should be fixed in commit b48aa0aba328f3bf6bb2741904c607a011c55292

magicDGS commented:
Fixed hard-tabs in commit 1572366678cd51a5ea64eb8c31b888be95331a26

tfenne commented on src/java/htsjdk/samtools/util/SamLocusIterator.java:
```diff
@@ -75,12 +77,15 @@ public RecordAndOffset(final SAMRecord record, final int offset) {
      * The unit of iteration.  Holds information about the locus (the SAMSequenceRecord and 1-based position
-     * on the reference), plus List of ReadAndOffset objects, one for each read that overlaps the locus
+     * on the reference), plus List of ReadAndOffset objects, one for each read that overlaps the locus;
+     * two more List of ReadAndOffset objects includes reads that overlaps the locus with insertions/deletions
```
this doesn't read quite right to me.  How about:
two more List_s_ of ReadAndOffset objects include reads that overlap the locus with insertions and deletions respectively.

magicDGS replied:
Done in ac5c8b5360ae8fca32325ae304578bd0929736dc

tfenne commented on src/java/htsjdk/samtools/util/SamLocusIterator.java:
```diff
@@ -75,12 +77,15 @@ public RecordAndOffset(final SAMRecord record, final int offset) {
         private final SAMSequenceRecord referenceSequence;
         private final int position;
         private final List<RecordAndOffset> recordAndOffsets = new ArrayList<RecordAndOffset>(100);
+        private final List<RecordAndOffset> deletedInRecord = new ArrayList<RecordAndOffset>();
```
This class is already quite slow due to all the object creation involved.  What do you think about creating these two ArrayList objects lazily at the first call to `addDeleted` and `addInsertion` since they may frequently be empty?

magicDGS replied:
Done in ac5c8b5360ae8fca32325ae304578bd0929736dc

tfenne commented on src/java/htsjdk/samtools/util/SamLocusIterator.java:
```diff
@@ -75,12 +77,15 @@ public RecordAndOffset(final SAMRecord record, final int offset) {
         private final int position;
         private final List<RecordAndOffset> recordAndOffsets = new ArrayList<RecordAndOffset>(100);
+        private final List<RecordAndOffset> deletedInRecord = new ArrayList<RecordAndOffset>();
+        private final List<RecordAndOffset> insertionInRecord = new ArrayList<RecordAndOffset>();
```
The two new names, and their methods, should have matching names.  Either `deletedInRecord` and `insertedInRecord` or `deletionInRecord` and `insertionInRecord`.  Personally I prefer the first pair.

magicDGS replied:
Done in ac5c8b5360ae8fca32325ae304578bd0929736dc

tfenne commented on src/java/htsjdk/samtools/util/SamLocusIterator.java:
```diff
@@ -353,9 +395,17 @@ private void accumulateSamRecord(final SAMRecord rec) {
         final int alignmentEnd    = rec.getAlignmentEnd();
         final int alignmentLength = alignmentEnd - alignmentStart;
 
+        // get the offset for an insertion if we are tracking them
```
Now that it's a little more complicated, could you please extract lines ~398-408 into a separate method called something like `ensureAccumulatorIsPopulated()`?

magicDGS replied:
I extracted the computation of the offset for the queue in  the method `getAccumulatorOffset` in commit 7c9be2b1d221233cadb5180b50984e6760888b87

tfenne commented on src/tests/java/htsjdk/samtools/util/SamLocusIteratorTest.java:
```diff
@@ -64,6 +64,34 @@ public void testBasicIterator() {
+    @Test
+    public void testBasicIteratorWithIndels() {
+
+        final String sqHeader = "@HD\tSO:coordinate\tVN:1.0\n@SQ\tSN:chrM\tAS:HG18\tLN:100000\n";
```
I feel bad making this comment, because this is totally not your fault.  You've made changes in a class that uses an outdated way of building SAMRecords for testing, and followed the pattern you found.  But I'd really prefer not to add lots more tests that do things this way.  Would you mind taking a look at `SAMRecordSetBuilder` and seeing if you could convert the tests to use that instead?

magicDGS replied:
I've just followed the pattern found to keep the class without many changes, but I could have a look to `SAMRecordSetBuilder` and use it instead. In addition, I also found that some of the `Asser.assertEquals()` from the original class are in the wrong order (expected/actual instead of actual/expected). Should I change that too?

tfenne replied:
If you're willing that would be great.

magicDGS replied:
Done in dbf3c75b170adda515385f52e6302701fd5cde38

tfenne commented on src/tests/java/htsjdk/samtools/util/SamLocusIteratorTest.java:
```diff
@@ -64,6 +64,34 @@ public void testBasicIterator() {
+    }
+
+    @Test
+    public void testBasicIteratorWithIndels() {
```
`testIteratorWithIndelsOnButNoIndels`?

magicDGS replied:
Added to `testBasicIterator` in dbf3c75b170adda515385f52e6302701fd5cde38

tfenne commented on src/tests/java/htsjdk/samtools/util/SamLocusIteratorTest.java:
```diff
@@ -93,6 +121,43 @@ public void testEmitUncoveredLoci() {
+    }
+
+    @Test
+    public void testEmitUncoveredLociWithIndels() {
```
Instead of copy/pasting the whole method with an additional call to `setIncludeIndels(true)` you could just wrap the body of the original method with `for (final boolean incIndels : new boolean[][ {true, false}) {`

magicDGS replied:
Sure, it will also reduce the number of methods (I will do it also with `testBasicIterator` to avoid the suggested `testIteratorWithIndelsOnButNoIndels`) in the previous comment.

magicDGS replied:
Done in dbf3c75b170adda515385f52e6302701fd5cde38

tfenne commented on src/tests/java/htsjdk/samtools/util/SamLocusIteratorTest.java:
```diff
@@ -119,6 +184,37 @@ public void testQualityFilter() {
+    }
+
+    @Test
+    public void testQualityFilterWithIndels() {
```
Same thing here about not duplicating the test code.

magicDGS replied:
Done in dbf3c75b170adda515385f52e6302701fd5cde38

tfenne commented:
Hi @magicDGS there's one case that I think doesn't work if I've understood the code correctly.  Either way it would be great to add a test for it to ensure/show that it works.  My concern is that I think if you encounter a stream of record with the same alignment start position, and the _first_ read starts with an `M` operator and some subsequent read starts with an `I` operator, things will go badly.  E.g.

```
R1	65	chr4	5000	60	101M	...
R2	65	chr4	5000	60	6I95M	...
R3	65	chr4	5000	60	101M	...
R4	65	chr4	5000	60	101M	...
```

I think in this case that when `R1` is encountered that the accumulator will be flushed up to position `5000`, and then all the bases from `R1` accumulated.  And then when `R2` is encountered, it will expect that `accumulator[0]` is in fact for position 4999, and the exception on line 349 will be thrown.

Since you can't rely on any specific ordering of reads with the same alignment start position, you may need to change things further such that you pull off all the reads with the same alignment start position, order them by whether or not they start with an indel, and then accumulate from all of them.

tfenne commented on src/java/htsjdk/samtools/util/SamLocusIterator.java:
```diff
@@ -384,6 +434,56 @@ private void accumulateSamRecord(final SAMRecord rec) {
+        // iterate over the cigar element
+        for (int elementIndex = 0; elementIndex < cigar.size(); elementIndex++) {
+            final CigarElement e = cigar.get(elementIndex);
+            switch (e.getOperator()) {
```
This switch statements would, I think, be clearer written as follows:
```java
if (operator == I) // handle insertion
else if (operator == D) // handle deletion
else {
    if (operator.consumesReadBases()) readBase += e.getLength();
    if (operator.consumesReferenceBases()) refBase += e.getLength();
}
```

magicDGS replied:
Done in 5d23bdd2f9fca3027257e060a3dd8741678eaf2c

tfenne commented:
@magicDGS I've made a first pass through the code, but will likely have more comments once you've addressed the first batch.

@alecw Would you mind taking a look at this PR, perhaps after @magicDGS has made the first round of revisions?  As the original author of `SamLocusIterator` I think it would be good for you to take a look at this.

magicDGS commented:
Hi @tfenne, it is true that the iterator is not working on the test case that you mention and I will modify the `testStartWithInsertion` method to develop what you suggest. Thank you very much for your revision, and I will let you know when I fix that issue.

magicDGS commented:
Hello @tfenne, for a reason that I do not fully understand, `testEmitUncoveredLoci` is broken when using the `SAMRecordSetBuilder` (commit 9d14bb9fa74a6fbd375515cd9ced0a61c442d4c2, see the file [here](https://github.com/samtools/htsjdk/blob/9d14bb9fa74a6fbd375515cd9ced0a61c442d4c2/src/tests/java/htsjdk/samtools/util/SamLocusIteratorTest.java)), but the implementation is the same as in the commit where the test pass (5d23bdd2f9fca3027257e060a3dd8741678eaf2c, see the file [here](https://github.com/magicDGS/htsjdk/blob/5d23bdd2f9fca3027257e060a3dd8741678eaf2c/src/tests/java/htsjdk/samtools/util/SamLocusIteratorTest.java)), although separating the `setIncludeIndels`.

Do you have any clue about what is happening?

alecw commented:
Truth be told, I'm not the original author.  We inherited this code from @kcibul.  Anyway, this code is pretty complicated (apart from the changes) and I don't think I have the time to get my head into this enough to comment intelligently. 

One suggestion I would make, however, is that it would be good for the javadoc to explain how insertions and deletions will be handled differently according to the value of includeIndels. As it stands now I don't know what the existing behavior is and what the new behavior will be.

-Alec

magicDGS commented:
Although I'm still working on this, I found that GATK4 already implemented a `LocusIteratorByState` (https://github.com/broadinstitute/gatk/pull/1442) that solve the initial issue.

I do not know if it is worth to implement this option in htsjdk or if someone is interested in that behavior use directly the GATK4 framework. What do you think, @tfenne?

droazen commented:
@tfenne Are you still interested in getting this into htsjdk? It sounds like @magicDGS has found a solution in GATK4 for the original need that prompted this PR. If you are interested, could you comment before the next htsjdk review party on 5/3? Otherwise, we might close this due to lack of interest.

magicDGS commented:
I found the solution in GATK4, but I think that it is still important in htsjdk. But it will probably be better for the new API if it is implemented (#520), using `AlignmentContext` as in GATK.

yfarjoun commented on src/java/htsjdk/samtools/util/SamLocusIterator.java:
```diff
@@ -287,8 +333,14 @@ public LocusInfo next() {
+            int start = rec.getAlignmentStart();
+            // only if we are including indels and the record does not start in the first base of the reference
+            // the stop locus to populate the queue is not the same if the record starts with an insertion
+            if(includeIndels && start != 1 && rec.getCigar().getCigarElement(0).getOperator().equals(CigarOperator.I)) {
```
what if there's a clip and then an insertion, e.g 5S6I4M . In this case, you'll not notice that it is in-fact an insertion that comes before the alignment start. (needs a test)

magicDGS replied:
Added test. Bug fix soon.

yfarjoun commented on src/java/htsjdk/samtools/util/SamLocusIterator.java:
```diff
@@ -431,8 +515,9 @@ else if (nextbit >= stopBeforeLocus.getPosition()) {
-        while (!accumulator.isEmpty() && accumulator.get(0).getRecordAndPositions().isEmpty() &&
+        // Skip over these if not including indels
+        while (!accumulator.isEmpty() && accumulator.get(0).isEmpty() &&
+                // accumulator.get(0).getRecordAndPositions().isEmpty() && !(includeIndels && !accumulator.get(0).getDeletedInRecord().isEmpty()) &&
```
why is this code commented? should it be removed?

magicDGS replied:
Removed.

yfarjoun commented on src/java/htsjdk/samtools/util/SamLocusIterator.java:
```diff
@@ -469,6 +554,29 @@ private void populateCompleteQueue(final Locus stopBeforeLocus) {
+        final int alignmentEnd    = rec.getAlignmentEnd();
+        final int alignmentLength = alignmentEnd - alignmentStart;
+        // get the offset for an insertion if we are tracking them
+        final int insOffset =
```
same issue as before regarding softclips.

magicDGS replied:
Done.

yfarjoun commented on src/java/htsjdk/samtools/util/SamLocusIterator.java:
```diff
@@ -469,6 +554,29 @@ private void populateCompleteQueue(final Locus stopBeforeLocus) {
+        final int insOffset =
+                (includeIndels && rec.getCigar().getCigarElement(0).getOperator().equals(CigarOperator.I)) ? 1 : 0;
+        // if there is an insertion in the first base and it is not tracked in the accumulator, add it
+        if(insOffset == 1 && accumulator.isEmpty()) {
```
```code
if (   
```
(needs a space)

magicDGS replied:
Done.

yfarjoun commented on src/tests/java/htsjdk/samtools/util/SamLocusIteratorTest.java:
```diff
@@ -161,10 +196,91 @@ public void testSimpleGappedAlignment() {
+     * Try all CIGAR operands (except H and P) and confirm that loci produced by SamLocusIterator are as expected with indels
+     */
+    @Test
+    public void testSimpleGappedAlignmentWithIndels() {
```
would be better to include this in a loop over `includeIndels` as you did in other tests.

magicDGS replied:
Done.

magicDGS commented:
I think that I fixed the bugs. Back to @yfarjoun.

magicDGS commented:
@akiezun, could you have a look to the failed test when both `testSimpleGappedAlignment()` were merged together? I don't understand why it is failing (it could be possible that it was the reason why I have them separate).

yfarjoun commented:
I think that this is a victim of the switch to gradle. Can you rebase this (without squashing) on top of master?

magicDGS commented:
It is already rebased, @yfarjoun, and I force the push with the squashed commits (so I can't retrieve the previous test). I can implement separate tests again from scratch to check if it is still not working.

yfarjoun commented:
In case you do not know: if you really want to recover your unsquashed
commits, you can look in `git reflog`..

On Tue, May 17, 2016 at 9:25 AM, Daniel Gómez-Sánchez <
notifications@github.com> wrote:

> It is already rebased, @yfarjoun <https://github.com/yfarjoun>, and I
> force the push with the squashed commits (so I can't retrieve the previous
> test). I can implement separate tests again from scratch to check if it is
> still not working.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/samtools/htsjdk/pull/408#issuecomment-219716273>
>

yfarjoun commented:
@lbergelson @droazen I think that this is due to the move to gradle...could you take a look?

magicDGS commented:
Thanks @yfarjoun, I didn't know and it will be useful. I think that there is a bug and not a gradle problem, but let's see what they think about.

magicDGS commented:
@lbergelson @droazen, it is not a problem of gradle. There is a bug in the code when accumulating deletions in a complicated case. I'll work on it.

yfarjoun commented:
Thanks Daniel.

On Tue, May 17, 2016 at 10:11 AM, Daniel Gómez-Sánchez <
notifications@github.com> wrote:

> @lbergelson <https://github.com/lbergelson> @droazen
> <https://github.com/droazen>, it is not a problem of gradle. There is a
> bug in the code when accumulating deletions in a complicated case. I'll
> work on it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/samtools/htsjdk/pull/408#issuecomment-219729642>
>

magicDGS commented:
@yfarjoun, I reorganized commits and squashed them for an easier review. I solved the bug in the code when tracking indels, and added some simple tests for them. Could you have a look?

yfarjoun commented on src/main/java/htsjdk/samtools/util/SamLocusIterator.java:
```diff
@@ -369,19 +428,29 @@ private boolean surpassedAccumulationThreshold() {
+        for(final CigarElement element: cigar.getCigarElements()) {
+            switch(element.getOperator()) {
+                case I: return true;
+                case S: continue;
```
this should be any operator that doesn't consume refbases (except "I"). I think it would be better to implement this as an if statement. Perhaps something like: 
```
<SNIP>
if( element.getOperator()==CigarOperator.I ) return true;
if ( ! element.getOperator().consumesReferenceBases() )  continue;
break;
</SNIP>
```

yfarjoun commented on src/main/java/htsjdk/samtools/util/SamLocusIterator.java:
```diff
@@ -369,19 +428,29 @@ private boolean surpassedAccumulationThreshold() {
     }
 
     /**
+     * Check if cigar start with an insertion, ignoring softclip
```
...ignoring other operators that do not consume references bases

yfarjoun commented on src/main/java/htsjdk/samtools/util/SamLocusIterator.java:
```diff
@@ -369,19 +428,29 @@ private boolean surpassedAccumulationThreshold() {
     /**
+     * Check if cigar start with an insertion, ignoring softclip
+     * @param cigar the cigar element
+     * @return <code>true</code> if it start with an insertion; <code>false</code> otherwise
```
...if the first operator to consume reference bases or be an insertion, is an insertion...

yfarjoun commented on src/main/java/htsjdk/samtools/util/SamLocusIterator.java:
```diff
@@ -369,19 +428,29 @@ private boolean surpassedAccumulationThreshold() {
 
     /**
+     * Check if cigar start with an insertion, ignoring softclip
+     * @param cigar the cigar element
```
remove the word element.

yfarjoun commented on src/main/java/htsjdk/samtools/util/SamLocusIterator.java:
```diff
@@ -300,8 +348,14 @@ public LocusInfo next() {
+            int start = rec.getAlignmentStart();
+            // only if we are including indels and the record does not start in the first base of the reference
+            // the stop locus to populate the queue is not the same if the record starts with an insertion
+            if(includeIndels && start != 1 && startWithInsertion(rec.getCigar())) {
```
```
if (
```
(space)

yfarjoun commented on src/main/java/htsjdk/samtools/util/SamLocusIterator.java:
```diff
@@ -494,6 +598,29 @@ private void populateCompleteQueue(final Locus stopBeforeLocus) {
+        }
+
+        // Ensure there are LocusInfos up to and including this position
+        for (int i=accumulator.size(); i<=alignmentLength+insOffset; ++i) {
```
spaces around operators

yfarjoun commented on src/test/java/htsjdk/samtools/util/SamLocusIteratorTest.java:
```diff
@@ -23,128 +23,327 @@
+        final int startPosition = 165;
+        for(int i = 0; i < coverage; i++) {
+            // add a negative-strand fragment mapped on chrM with base quality of 10
+            builder.addFrag("record"+i, 0, startPosition, true, false, "36M", null, 10);
```
spaces

yfarjoun commented on src/test/java/htsjdk/samtools/util/SamLocusIteratorTest.java:
```diff
@@ -23,128 +23,327 @@
+        for(int i = 0; i < coverage; i++) {
+            final String qualityString;
+            // half of the reads have a different quality
+            if(i % 2 == 0) {
```
``` 
if (
```

yfarjoun commented on src/test/java/htsjdk/samtools/util/SamLocusIteratorTest.java:
```diff
@@ -23,128 +23,327 @@
+                qualityString = "+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*+*";
+            }
+            // add a negative-strand fragment mapped on chrM with base quality of 10
+            builder.addFrag("record"+i, 0, startPosition, true, false, "36M", qualityString, 10);
```
spaces

yfarjoun commented on src/test/java/htsjdk/samtools/util/SamLocusIteratorTest.java:
```diff
@@ -23,128 +23,327 @@
+        final int startPosition = 165;
+        for(int i = 0; i < coverage; i++) {
+            // add a negative-strand fragment mapped on chrM with base quality of 10
+            builder.addFrag("record"+i, 0, startPosition, true, false, "18M10D18M", null, 10);
```
spaces

yfarjoun commented on src/test/java/htsjdk/samtools/util/SamLocusIteratorTest.java:
```diff
@@ -23,128 +23,327 @@
+                    isDeletedPosition = false;
+                }
+                Assert.assertEquals(li.getPosition(), pos++);
+                if(isDeletedPosition) {
```
```
if (
```

yfarjoun commented on src/test/java/htsjdk/samtools/util/SamLocusIteratorTest.java:
```diff
@@ -23,128 +23,327 @@
+                Assert.assertEquals(li.getRecordAndPositions().size(), coverage);
+                // make sure that we are not accumulating deletions
+                Assert.assertEquals(li.getDeletedInRecord().size(), 0);
+                if(incIndels && li.getPosition() == insStart) {
```
```
if (
```

yfarjoun commented on src/test/java/htsjdk/samtools/util/SamLocusIteratorTest.java:
```diff
@@ -23,128 +23,327 @@
+        final SAMRecordSetBuilder builder = getRecordBuilder();
+        // add records up to coverage for the test in that position
+        final int startPosition = 165;
+        for(int i = 0; i < coverage; i++) {
```
```
for (
```

yfarjoun commented on src/test/java/htsjdk/samtools/util/SamLocusIteratorTest.java:
```diff
@@ -23,128 +23,327 @@
+            int pos = startPosition;
+            for (final SamLocusIterator.LocusInfo li : sli) {
+                boolean isDeletedPosition = (pos >= deletionStart && pos <= deletionEnd);
+                if(!incIndels && isDeletedPosition) {
```
```
if (
```

yfarjoun commented on src/test/java/htsjdk/samtools/util/SamLocusIteratorTest.java:
```diff
@@ -23,128 +23,327 @@
+        final int deletionStart = 183;
+        final int deletionEnd = 192;
+        // test both for include indels and do not include indels
+        for (final boolean incIndels : new boolean[]{false, true}){
```
```
[] {false, true}) {
```

yfarjoun commented on src/test/java/htsjdk/samtools/util/SamLocusIteratorTest.java:
```diff
@@ -23,128 +23,327 @@
+        final SAMRecordSetBuilder builder = getRecordBuilder();
+        // add records up to coverage for the test in that position
+        final int startPosition = 165;
+        for(int i = 0; i < coverage; i++) {
```
space

yfarjoun commented:
Thanks for your work on this PR @magicDGS.

I've commented about your bug fix, and a few of the places where the spaces in your code are missing (sorry for the nit-picking, we are trying to have a consistent code style). There are more that I missed, perhaps you can set your IDE to fix these for you?

I think that there are some CIGAR cases that are missing. 

e.g.

```
2M4N3I4M
2M4N4D5M (is this legal?)
```

magicDGS commented:
Back to you for (final?) review, @yfarjoun:

- [x] Coding style fixed with IDE
- [x] `startWithInsertion()` implemented as you suggested
- [x] New tests for examples of CIGAR in previous comment

yfarjoun commented:
Thanks @magicDGS I'm happy with this. 👍 

@tfenne you took a look at this first, and said you might have more comments. Are you happy enough with my review, or would you like another glance?

yfarjoun commented:
Thanks @magicDGS !!
